test/e2e/framework/conformance/conformance.go

```diff
@@ -0,0 +1,464 @@
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package architecture
+
+import (
+	"bytes"
+	"context"
+	"fmt"
+	"slices"
+	"strings"
+	"time"
+
+	"github.com/google/go-cmp/cmp"
+	"github.com/onsi/ginkgo/v2"
+	"github.com/onsi/gomega"
+
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	"k8s.io/apimachinery/pkg/types"
+	"k8s.io/client-go/dynamic"
+	"k8s.io/client-go/rest"
+	"k8s.io/klog/v2"
+	"k8s.io/kubernetes/test/e2e/framework"
+	"sigs.k8s.io/yaml"
+)
+
+// ResourceTestcase describes how to test one particular API endpoint
+// by executing different operations against it. Each testcase
+// automatically becomes part of the conformance test
+// suite, so only endpoints for GA features may be listed here.
+//
+// This checks basic create/read/update/delete (CRUD) semantic, which
+// is the minimum that is required for conformance testing of a
+// GA feature. Actual functional testing is desirable, but not
+// required.
+type ResourceTestcase struct {
+	// APIVersion contains the API group and version using the format <group>/<version>.
+	APIVersion string
+
+	// Resource is the name of the Resource within the API (plural form, lower case).
+	Resource string
+
+	// Namespaced must be true if the resource must be created in a
+	// namespace, false if it is cluster-scoped.
+	//
+	// Namespaced resources get created in the test namespace.
+	// The name of cluster-scoped resources gets extended with
+	// `-<test namespace name>` to make it unique.
+	Namespaced bool
+
+	// HasStatus must be set to true to enable testing of the
+	// status sub-resource.
+	HasStatus bool
+
+	// InitialData must contain the JSON or YAML representation
+	// of a valid resource, including a status if the resource
+	// has one.
+	InitialData string
+
+	// PatchData must contain modified fields for spec and
+	// (if applicable) status for use in a strategic merge patch.
+	// Must include apiVersion and kind. May contain the name,
+	// but that's not required.
+	PatchData string
+}
+
+// TestResource covers all the typical endpoints for a resource through
+// dynamic client calls.
+func TestResource(ctx context.Context, f *framework.Framework, tc ResourceTestcase) {
+	// Set up clients.
+	gv, err := schema.ParseGroupVersion(tc.APIVersion)
+	framework.ExpectNoError(err, "parse apiVersion")
+	gvr := gv.WithResource(tc.Resource)
+	resourceClient := f.DynamicClient.Resource(gvr)
+	var client dynamic.ResourceInterface
+	if tc.Namespaced {
+		client = resourceClient.Namespace(f.Namespace.Name)
+	} else {
+		client = resourceClient
+	}
+	config := dynamic.ConfigFor(f.ClientConfig())
+	httpClient, err := rest.HTTPClientFor(config)
+	framework.ExpectNoError(err, "construct HTTP client")
+	restClient, err := rest.UnversionedRESTClientForConfigAndClient(config, httpClient)
```

Comment:
do we have to use a raw REST client directly? can we use a discovery client to fetch discovery endpoints for this?

Reply:
this is fine, I see us directly requesting discovery endpoints in other places